### PR TITLE
refactor(api): incorporate module core into non-engine PAPI

### PR DESCRIPTION
**Pull request by mcous**
## Overview

This PR takes the module interfaces added in #11437 and partially integrates them into the existing PAPIv2 module contexts. As of this PR, all module contexts have a `_core` that they delegate to for labware loading.

**This PR concentrates on the existing PAPIv2 cores**. It does not implement anything for the new engine-based core.

The remaining work after this PR is to :

- Implement `load_module` on the `ProtocolEngine`-based protocol core
- Implement `load_labware` for labware-on-modules on the PE-based common module core
- Create per-module-type module cores that will:
    - Engine-based cores: dispatch `ProtocolEngine` module commands
    - Legacy cores: call methods of the `SynchronousAdapter` module hardware API wrappers

## Review requests

I have smoke tested PAPIv2 module loading using the current PAPIv2 core logic on this branch. I smoke tested all modules, including a thermocycler in the `semi` config

- Module loads
- A module command works
- Labware on module loads
- Labware well access works

Test results

- [x] Magnetic module
- [x] Temperature module
- [x] Heater-shaker module
- [x] Thermocylcer module
- [x] Thermocycler module in `semi` config
    - Note: thermocycler `semi` config is broken in various ways in `edge` (RSS-105, RSS-106)
    - This branch behaves exactly the same as `edge`

## Risk assessment

High, this is messing about in production module code. Mitigations:

- Extensive smoke testing of real module hardware
- G-code regression suites are passing (and caught one bug while this PR was in development)

### CODE REVIEW

**Comment by mcous** [comment]
Current dependencies on/for this PR:
* edge
  * **PR #11443** <a href="https://app.graphite.dev/github/pr/Opentrons/opentrons/11443" target="_blank">Graphite</a> 
    * **PR #11437** <a href="https://app.graphite.dev/github/pr/Opentrons/opentrons/11437" target="_blank">Graphite</a> 
      * **PR #11449** <a href="https://app.graphite.dev/github/pr/Opentrons/opentrons/11449" target="_blank">Graphite</a>  👈

This comment was auto-generated by [Graphite](https://app.graphite.dev/github/pr/Opentrons/opentrons/11449?utm_source=stack-comment).

**Comment by mcous** [comment]
Current dependencies on/for this PR:
* edge
  * **PR #11443** <a href="https://app.graphite.dev/github/pr/Opentrons/opentrons/11443" target="_blank">Graphite</a> 
    * **PR #11437** <a href="https://app.graphite.dev/github/pr/Opentrons/opentrons/11437" target="_blank">Graphite</a> 
      * **PR #11449** <a href="https://app.graphite.dev/github/pr/Opentrons/opentrons/11449" target="_blank">Graphite</a>  👈

This comment was auto-generated by [Graphite](https://app.graphite.dev/github/pr/Opentrons/opentrons/11449?utm_source=stack-comment).

**Comment by mcous** [comment]
Current dependencies on/for this PR:
* edge
  * **PR #11443** <a href="https://app.graphite.dev/github/pr/Opentrons/opentrons/11443" target="_blank">Graphite</a> 
    * **PR #11437** <a href="https://app.graphite.dev/github/pr/Opentrons/opentrons/11437" target="_blank">Graphite</a> 
      * **PR #11449** <a href="https://app.graphite.dev/github/pr/Opentrons/opentrons/11449" target="_blank">Graphite</a>  👈

This comment was auto-generated by [Graphite](https://app.graphite.dev/github/pr/Opentrons/opentrons/11449?utm_source=stack-comment).

**Comment by mcous** [comment]
Current dependencies on/for this PR:
* edge
  * **PR #11443** <a href="https://app.graphite.dev/github/pr/Opentrons/opentrons/11443" target="_blank">Graphite</a> 
    * **PR #11437** <a href="https://app.graphite.dev/github/pr/Opentrons/opentrons/11437" target="_blank">Graphite</a> 
      * **PR #11449** <a href="https://app.graphite.dev/github/pr/Opentrons/opentrons/11449" target="_blank">Graphite</a>  👈

This comment was auto-generated by [Graphite](https://app.graphite.dev/github/pr/Opentrons/opentrons/11449?utm_source=stack-comment).

**Comment by mcous** [comment]
Current dependencies on/for this PR:
* edge
  * **PR #11443** <a href="https://app.graphite.dev/github/pr/Opentrons/opentrons/11443" target="_blank">Graphite</a> 
    * **PR #11437** <a href="https://app.graphite.dev/github/pr/Opentrons/opentrons/11437" target="_blank">Graphite</a> 
      * **PR #11449** <a href="https://app.graphite.dev/github/pr/Opentrons/opentrons/11449" target="_blank">Graphite</a>  👈

This comment was auto-generated by [Graphite](https://app.graphite.dev/github/pr/Opentrons/opentrons/11449?utm_source=stack-comment).

File: api/src/opentrons/protocol_api/core/engine/protocol.py [hunk 1]
```diff
@@ -84,12 +84,15 @@ def add_labware_definition(
     def load_labware(
         self,
         load_name: str,
-        location: DeckSlotName,
+        location: Union[DeckSlotName, ModuleCore],
```
Comment (TamarZanzouri) [inline api/src/opentrons/protocol_api/core/engine/protocol.py:87]: nice! took me a moment to understand this additional type

File: api/src/opentrons/protocol_api/core/engine/protocol.py [hunk 2]
```diff
@@ -107,14 +110,10 @@ def load_module(
         model: ModuleModel,
         location: Optional[DeckSlotName],
         configuration: Optional[str],
-    ) -> Optional[LoadModuleResult]:
+    ) -> ModuleCore:
         """Load a module into the protocol."""
         raise NotImplementedError("ProtocolEngine PAPI core not implemented")
 
-    def get_loaded_modules(self) -> Dict[int, LoadModuleResult]:
```
Comment (TamarZanzouri) [inline api/src/opentrons/protocol_api/core/engine/protocol.py:114]: is this not being used anymore?

Comment (mcous) [reply api/src/opentrons/protocol_api/core/engine/protocol.py:114]: Yes, I found I could replace it with a simple dictionary in the top level `ProtocolContext` class. I couldn't find a compelling reason for this logic to live in the implementation core rather than the top-level, given that it's common logic across all implementations

**Review by jbleon95 — APPROVED** [review]
Looks good to me, in line with everything we've done so far

File: api/src/opentrons/protocol_api/protocol_context.py [hunk 3]
```diff
@@ -686,3 +672,26 @@ def rail_lights_on(self) -> bool:
     def door_closed(self) -> bool:
         """Returns True if the robot door is closed"""
         return self._implementation.door_closed()
+
+
+def _create_module_context(
+    core: ModuleCore,
```
Comment (TamarZanzouri) [inline api/src/opentrons/protocol_api/protocol_context.py]: can we change the name of this argument? `module_core` would be enough if that works

File: api/src/opentrons/protocols/geometry/deck.py [hunk 4]
```diff
@@ -53,8 +53,9 @@ def __init__(self) -> None:
             self.data[int(slot["id"])] = None
             self._positions[int(slot["id"])] = types.Point(*slot["position"])
         self._highest_z = 0.0
-        self._load_fixtures()
         self._thermocycler_present = False
+        self._load_fixtures()
+        self.recalculate_high_z()
```
Comment (TamarZanzouri) [inline api/src/opentrons/protocols/geometry/deck.py]: why was this added?

File: api/tests/opentrons/protocol_api/core/legacy/test_protocol_context_implementation.py [hunk 5]
```diff
@@ -204,3 +206,87 @@ def test_load_labware(
         ),
         times=1,
     )
+
+
+def test_load_labware_on_module(
```
Comment (TamarZanzouri) [inline api/tests/opentrons/protocol_api/core/legacy/test_protocol_context_implementation.py:211]: Nice!

**Comment by TamarZanzouri** [comment]
Changes look great to me! Had a bunch of questions. Just a bit confused with the use of the same file names for _core and non _core. made this a bit harder for me to follow. Sorry 😞

File: api/src/opentrons/protocols/geometry/deck.py (hunk 4 above)
Comment (mcous) [reply api/src/opentrons/protocols/geometry/deck.py]: Oh oops, good catch; it's not needed. My changes introduced a z-height bug and I was searching around for a fix when I added this

File: api/src/opentrons/protocol_api/protocol_context.py (hunk 3 above)
Comment (mcous) [reply api/src/opentrons/protocol_api/protocol_context.py]: It's a little hard to tell in these PRs, so I'm sorry about that, but in general when you're reading through these:

- `core` means "core control logic"
- `core` means the same as `implementation`
- I will rename all the `implementation` fields / classes to `core` before this is all over

Comment (TamarZanzouri) [reply api/src/opentrons/protocol_api/protocol_context.py]: Thank you for explaining that!

**Comment by mcous** [comment]
**Graphite Merge Job**

Current status: ✅ Merged

This pull request was successfully merged as part of a stack.

This comment was auto-generated by [Graphite](https://app.graphite.dev/github/pr/Opentrons/opentrons/11449?utm_source=stack-comment).


Job Reference: VI231HKmSMYZgESfS8Oo
